### PR TITLE
Fix a crash when `# fmt: on` is used on a different block level than `# fmt: off`

### LINKED ISSUE #569: Invalid code after indented `# fmt: off`
Input:
```py3
def x():
    # fmt: off
    return

def y():
    return
```

Invalid code produced:
```py3
def x():
    # fmt: off
    return

    def y():
        return
```

### LINKED ISSUE #1817: Unhelpful error message if you forget `# fmt: on` after a `# fmt: off`
**Describe the bug** 

If you forget to add `# fmt: on` after a `# fmt: off`, you get an unhelpful error message. 

**To Reproduce** 

1. Take this file `repro.py`
```python
def f():
    # fmt: off
    pass

if True:
    pass
```

2. Run `black --check repro.py`

3. See error:
```
error: cannot format repro.py: INTERNAL ERROR: Black produced code that is not equivalent to the source.  Please report a bug on https://github.com/psf/black/issues.  This diff might be helpful: /tmp/blk_k081_p1l.log
Oh no! 💥 💔 💥
1 file would fail to reformat.
```

**Expected behavior** 
I would expect that this either works (and it just ignores the rest of the file), or that you get an error message telling you that you need to add `# fmt: on` somewhere.

**Environment (please complete the following information):**

- Version:  20.8b1
- OS and Python version: Ubuntu 20.04, Python 3.8.5

**Does this bug also happen on master?**
Yes, it happens on  https://black.now.sh/?version=master

**Additional context**

Here's the output of `/tmp/blk_k081_p1l.log`:
```
--- src
+++ dst
@@ -14,26 +14,26 @@
             None,  # NoneType
         )  # /arguments
       body=
         Pass(
         )  # /Pass
+        If(
+          body=
+            Pass(
+            )  # /Pass
+          orelse=
+          test=
+            Constant(
+              value=
+                True,  # bool
+            )  # /Constant
+        )  # /If
       decorator_list=
       name=
         'f',  # str
       returns=
         None,  # NoneType
       type_comment=
         None,  # NoneType
     )  # /FunctionDef
-    If(
-      body=
-        Pass(
-        )  # /Pass
-      orelse=
-      test=
-        Constant(
-          value=
-            True,  # bool
-        )  # /Constant
-    )  # /If
   type_ignores=
 )  # /Module
```

### LINKED ISSUE #2140: fmt: off/on on different indentation levels leads to INTERNAL ERROR: Black produced invalid code on pass 1
Placing several `fmt: off`/`fmt: on` comments, each around long `with ... as ...:` lines, leads to black raising  an internal error

**To Reproduce** Steps to reproduce the behavior:

1. Take this file `foo.py`
    ```
    name = "foo.py"


    def foo():
        # fmt: off
        with open(name) as f1, open(name) as f2, open(name) as f3, open(name) as f4, open(name) as f5:
            # fmt: on
            print(f1, f2, f3, f4, f5)


    def bar():
        # fmt: off
        with open(name) as f1, open(name) as f2, open(name) as f3, open(name) as f4, open(name) as f5:
            # fmt: on
            print(f1, f2, f3, f4, f5)
    ```
2. Run _Black_ on it with these arguments: `black foo.py`
3. See error: `error: cannot format foo.py: INTERNAL ERROR: Black produced invalid code on pass 1: invalid syntax (<unknown>, line 12). Please report a bug on https://github.com/psf/black/issues.  This invalid output might be helpful: /tmp/blk_3i4qdouv.log`

**Expected behavior** Black formats the source file, leaving the line marked to not be formatted alone.

**Environment (please complete the following information):**

- Version: black, version 21.4b0
- OS and Python version: Linux, Python 3.9.4

**Does this bug also happen on master?** [Yes it does](https://black.vercel.app/?version=master&state=_Td6WFoAAATm1rRGAgAhARYAAAB0L-Wj4AHUAJldAD2IimZxl1N_WmFm63wsQYxa3iKd5NgD9wTuCPA8tXPZto6Z99Hjctu-G02vQ66iazqXJ-pZayASz_zt2d6hMht-Jz3nSIdAJx8Lbu-jjCl2N3lDI4xfh7bbsLR1KxCmyuhwYrFT9mXzxip7MhJ4QfBbWxURZlPZnaqHmrP9cGhxLd6aOZp7upBQiGacF0GSuolpk9M0NbREAAAAAACaKx9iKoxo3QABtQHVAwAAWYj-p7HEZ_sCAAAAAARZWg==)

**Additional context** Add any other context about the problem here.

The helpful diff:
<details>
<pre>
  File "/home/mkhl/.cache/pypoetry/virtualenvs/api-FV8o_s3F-py3.9/lib/python3.9/site-packages/black/__init__.py", line 6508, in assert_equivalent
    dst_ast = parse_ast(dst)
  File "/home/mkhl/.cache/pypoetry/virtualenvs/api-FV8o_s3F-py3.9/lib/python3.9/site-packages/black/__init__.py", line 6409, in parse_ast
    return ast27.parse(src)
  File "/home/mkhl/.cache/pypoetry/virtualenvs/api-FV8o_s3F-py3.9/lib64/python3.9/site-packages/typed_ast/ast27.py", line 50, in parse
    return _ast27.parse(source, filename, mode)
name = "foo.py"

def foo():
    # fmt: off
    with open(name) as f1, open(name) as f2, open(name) as f3, open(name) as f4, open(name) as f5:
        # fmt: on
        print(f1, f2, f3, f4, f5)

    def bar():
        # fmt: off
    with open(name) as f1, open(name) as f2, open(name) as f3, open(name) as f4, open(name) as f5:
        # fmt: on
        print(f1, f2, f3, f4, f5)

### LINKED ISSUE #2232: Crash with fmt: off + decorator
**Describe the bug** 

I am getting an exception `error: cannot format test_black.py: INTERNAL ERROR: Black produced code that is not equivalent to the source.` within a very specific scenario.

**To Reproduce**:

1. Take this file 'test_black.py'

```python
import pytest


class TestDummy:
    # fmt: off
    @pytest.mark.parametrize("param1", ["foo", "bar"])  # line A
    # fmt: on
    @pytest.mark.skip("special line")  # line B
    def test_dummy(self, param1):
        print(param1)


print("if you remove me, it will work too")  # line C

```

2. Run _Black_ on it

```shell
$ black test_black.py
```

3. See error

```
error: cannot format test_black.py: INTERNAL ERROR: Black produced code that is not equivalent to the source.  Please report a bug on https://github.com/psf/black/issues.  This diff might be helpful: /tmp/blk_s3rf0vdz.log
Oh no! 💥 💔 💥
1 file failed to reformat.
```

**Expected behavior** 

The file is already well-formatted. I was expecting no error on the console.

**Environment:**

- Version: 21.5b1  or   main
- OS and Python version: Linux with Python 3.8.5
- also on OS and Python version: macOS Darwin 19.6.0 with Python 3.8.0

**Does this bug also happen on main?**:

Yes, it happens.

**Additional context**

This file has a minimal setup to reproduce the error:

- if `# fmt: off` (above line with comment `# line A`) is removed  OR...
- if the line with the comment `# line B` is removed (only for main and for macOS)  OR...
- if the line with comment `# line C` is removed

then Black run without any error.

Here is the attached log file:
[blk_zpx9dsc4.log](https://github.com/psf/black/files/6476092/blk_zpx9dsc4.log)

### LINKED ISSUE #2567: Black messes up indentation of else block after for-statement in fmt on/off block
**Black https://github.com/psf/black/commit/62ed53**

The MRE below shows the issue pretty clearly, I think, but there's also a [Playground][playground] available.  
The [docs](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html) state:

> fmt: on/off have to be on the same level of indentation.

So this kind of single line block should be okay, shouldn't it? `#fmt: skip` didn't work for me, either.

For what it's worth, I came across this issue in a Docker container based on Ubuntu 20.04 with Python 3.8.10 and **black** v21.9b0. Only then, I tested against the latest main using the Playground.

## Options

`--line-length=88` -> this doesn't matter, I've only been using it in the Playground since it's a default there
`--safe`

## Input

```python

if True:
    #fmt: off
    for _ in range(1):
    #fmt: on
        print("This should get printed")
else:
    print("This shouldn't")

```

## Output

Please take a look at the [Playground][playground] for the entire output, it seemed too long to post here.

If run with `--fast`, the reformatted file becomes:

```python

if True:
    #fmt: off
    for _ in range(1):
    #fmt: on
        print("This should get printed")
    else:
        print("This shouldn't")

 ```

As you can see, the `else` block gets indented, which is wrong although it's still valid Python.

## Expected

```python

if True:
    #fmt: off
    for _ in range(1):
    #fmt: on
        print("This should get printed")
else:
    print("This shouldn't")

```

[playground]: https://black.vercel.app/?version=main&state=_Td6WFoAAATm1rRGAgAhARYAAAB0L-Wj4ADwAI1dAD2IimZxl1N_Wk6tt2q7tSLl4fLzSlExTOe62cV85aNDqCpOaxXFob5mQnBW7EOhxkJcp094SrFs5Y2mbZBhnBBDIRKeZ7WUPqswGCaSh1Kgwf-UL39x7qWPSbj4lZBrvDxO53d1Yue-ugjbuqO46w2QtMSNhO74FMXBSsks1cHu704LieJUtTZFcGrEgAAAAAAaiPeLMc0-VQABqQHxAQAAvqY-_7HEZ_sCAAAAAARZWg==

### LINKED ISSUE #2882: INTERNAL ERROR: Black produced code that is not equivalent to the source
**Describe the bug**

Error formatting python sources in https://github.com/pytorch/fairseq

**To Reproduce**

Clone [fairseq](https://github.com/pytorch/fairseq) and run black on offending files:

```
fairseq $ black --version
black, 22.1.0 (compiled: yes)

fairseq $ black --check \
  examples/simultaneous_translation/modules/monotonic_multihead_attention.py \
  examples/speech_text_joint_to_text/criterions/text_guide_cross_entropy_acc.py \
  examples/wav2vec/unsupervised/scripts/pca.py

error: cannot format examples/wav2vec/unsupervised/scripts/pca.py: INTERNAL ERROR: Black produced code that is not equivalent to the source.  Please report a bug on https://github.com/psf/black/issues.  This diff might be helpful: /var/folders/jg/hrl2p4ms1vjbn44gmmyl4p7r0000gn/T/blk_lda03pi7.log
error: cannot format examples/speech_text_joint_to_text/criterions/text_guide_cross_entropy_acc.py: INTERNAL ERROR: Black produced code that is not equivalent to the source.  Please report a bug on https://github.com/psf/black/issues.  This diff might be helpful: /var/folders/jg/hrl2p4ms1vjbn44gmmyl4p7r0000gn/T/blk_ctr_udzr.log
error: cannot format examples/simultaneous_translation/modules/monotonic_multihead_attention.py: INTERNAL ERROR: Black produced code that is not equivalent to the source.  Please report a bug on https://github.com/psf/black/issues.  This diff might be helpful: /var/folders/jg/hrl2p4ms1vjbn44gmmyl4p7r0000gn/T/blk_u26gqsix.log

Oh no! 💥 💔 💥
3 files would fail to reformat.
```

Debug diffs:

- [examples/simultaneous_translation/modules/monotonic_multihead_attention.py](https://gist.github.com/d0c11fed21612e54094d34b480b4cd20)
- [examples/speech_text_joint_to_text/criterions/text_guide_cross_entropy_acc.py](https://gist.github.com/8475c6128a4259d352b6c940edd9315e)
- [examples/wav2vec/unsupervised/scripts/pca.py](https://gist.github.com/f2c873f5b0c35133efffbb595fee5d2f)

**Environment**

- Black's version: 22.1.0
- OS and Python version: macOS 12.1

### LINKED ISSUE #2985: Fail gracefully when unaligned or missing 'fmt: off/on' found
The following snippet makes black fail with "Black produced code that is not equivalent to the source.".

```
import typing as t

class Named(t.Protocol):
    # fmt: off
    @property
    def title(self) -> str: ...

class Factory(t.Protocol):
    def __call__(self, **kwargs) -> Named: ...
    # fmt: on
```

The following works:

```
import typing as t

class Named(t.Protocol):
    # fmt: off
    @property
    def title(self) -> str: ...
    # fmt: on

class Factory(t.Protocol):
    # fmt: off
    def __call__(self, **kwargs) -> Named: ...
    # fmt: on
```
---

Side note: mistyping causes black to indicate the wrong error:

```
import typing as t

class Named(t.Protocol):
    # fmt: off
    @property
    def title(self) -> str: ...
    # fmt: one
    #        ^- "mistyped"

class Factory(t.Protocol):
    # fmt: off
    def __call__(self, **kwargs) -> Named: ...
    # fmt: on
```

In this case, black says "`Cannot parse: 13:4:     def __call__(self, **kwargs) -> Named: ...`"; which is misleading.

### LINKED ISSUE #3184: Black crashes when disabling fmt in some parts of the code
**Describe the bug**

Black crashes when disabling fmt in some parts of the code

**To Reproduce**

For example, take this code:

```python
class A:
    async def call(param):
        if param:
            # fmt: off
            if param[0:4] in (
                "ABCD", "EFGH"
            ):
                # fmt: on
                return 1

            elif param[0:4] in ("ZZZZ",):
                return 2

            elif param[0:4] in ("YYYY",):
                return 3

        return 4
```

If the fmt:off/on is moved to the elif, then it works:

```python
class A:
    async def call(param):
        if param:
            if param[0:4] in (
                "ABCD", "EFGH"
            ):
                return 1

            # fmt: off
            elif param[0:4] in (
                "ZZZZ",
            ):
                # fmt: on
                return 2

            elif param[0:4] in ("YYYY",):
                return 3

        return 4
```

The resulting error is:

> INTERNAL ERROR: Black produced code that is not equivalent to the source.  Please report a bug on https://github.com/psf/black/issues.  This diff might be helpful: /tmp/blk_ywpuqfap.log

**Expected behavior**

To format the code

**Environment**

Black (vercel): https://black.vercel.app/?version=main&state=_Td6WFoAAATm1rRGAgAhARYAAAB0L-Wj4AHUALtdAD2IimZxl1N_WlXnON2nzNApqtITJKNM5khekDYCHqoZycPfWzsPujeCAa9Rpjc9QNsUFfcNW21kfXWi3Ds1zEobtzgrlyJFv6pJgbr3hjVNgtP72qI38X2kZz6KEWMeQHyjVFNfYPRCWhXM_df-ORVEnAhEjzVHVfUOGnT6eg7CFQGFmjFuPqN9w6a87oJlA5IKn438pwNRwAkPWitXFDzX-irhc11HxGrYugCMb6BF-LvIjSN0daMSCgAAAFoGUhUf-a1_AAHXAdUDAAAIJsLiscRn-wIAAAAABFla

**Additional context**

### PR DESCRIPTION
**Pull request by yilei**
### Description

Previously _Black_ produces invalid code because the `# fmt: on` is used on a different block level.

While _Black_ requires `# fmt: off` and `# fmt: on` to be used at the same block level, incorrect usage shouldn't cause crashes.

The formatting behavior this PR introduces is, the code below the initial `# fmt: off` block level will be turned off for formatting, when `# fmt: on` is used on a different level or there is no `# fmt: on`. This also matches the current behavior when `# fmt: off` is used at the top-level without a matching `# fmt: on`, it turns off formatting for everything below `# fmt: off`.

- Fixes #2567
- Fixes #3184
- Fixes #2985
- Fixes #2882
- Fixes #2232
- Fixes #2140
- Fixes #1817
- Fixes #569

### Checklist - did you ...

- [x] Add a CHANGELOG entry if necessary?
- [x] Add / update tests if necessary?
- [x] Add new / update outdated documentation?

### CODE REVIEW

**Comment by yilei** [comment]
Now I think about it, this PR is effectively proposing the ability to use a single `# fmt: off` to turn off formatting by scope. Not sure about maintainers' opinion, but I think this improves ergonomics a lot. This also matches Pylint's `# pylint: disable=` pragma behavior, so it can be familiar to Pylint users. I do like to see this being supported.

You can already use a single `# fmt: off` at the top level to turn off formatting for everything below, though I don't see this explicitly documented.

**Comment by jack1142** [comment]
FYI, you need to repeat "fixes" keyword for each issue you're closing for them to get linked correctly:
https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

File: tests/data/simple_cases/fmtonoff5.py [hunk 1]
```diff
@@ -34,3 +34,125 @@ def test_func():
         return True
 
     return False
+
+
+# Regression test for https://github.com/psf/black/issues/2567.
+if True:
+    # fmt: off
+    for _ in range(1):
```
[the remark this answers is not in the document]
Comment (yilei) [reply tests/data/simple_cases/fmtonoff5.py]: Done

**Comment by yilei** [comment]
Thanks for the review, updated the test, please take another look.

**Pull request merged**